dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/queryengine/SourcesToStartingPoints.scala

```diff
@@ -9,48 +9,172 @@ import io.shiftleft.semanticcpg.language.operatorextension.allAssignmentTypes
 import scala.util.{Failure, Success, Try}
 
 case class StartingPointWithSource(startingPoint: CfgNode, source: StoredNode)
-
+case class UsageInput(src: StoredNode, typeDecl: TypeDecl, astNode: AstNode)
+case class ResultSummary(result: List[StartingPointWithSource], methodTasks: List[UsageInput])
 object SourcesToStartingPoints {
 
   private val log = LoggerFactory.getLogger(SourcesToStartingPoints.getClass)
 
   def sourceTravsToStartingPoints[NodeType](sourceTravs: IterableOnce[NodeType]*): List[StartingPointWithSource] = {
-    val fjp = ForkJoinPool.commonPool()
+    val executorService = Executors.newWorkStealingPool()
     try {
-      fjp.invoke(new SourceTravsToStartingPointsTask(sourceTravs: _*)).distinct
+      val sources = sourceTravs
+        .flatMap(_.iterator)
+        .collect { case n: StoredNode => n }
+        .dedup
+        .toList
+      sources.headOption
+        .map(src => {
+          // We need to get Cpg wrapper from graph. Hence we are taking head element from source iterator.
+          // This will also ensure if the source list is empty then these tasks are invoked.
+          val cpg                           = Cpg(src.graph())
+          val (startingPoints, methodTasks) = calculateStartingPoints(sources, executorService)
+          val startingPointFromUsageInOtherClasses =
+            calculateStatingPointsWithUsageInOtherClasses(methodTasks, cpg, executorService)
+          (startingPoints ++ startingPointFromUsageInOtherClasses)
+            .sortBy(_.source.id)
+        })
+        .getOrElse(Nil)
     } catch {
       case e: RejectedExecutionException =>
         log.error("Unable to execute 'SourceTravsToStartingPoints` task", e); List()
     } finally {
-      fjp.shutdown()
+      executorService.shutdown()
     }
   }
 
+  /** This will process and identify the starting points except the usage in other classes. This run will identify the
+    * required tasks for calculating starting points with usage in other classes.
+    *
+    * @param sources
+    *   \- Sources list
+    * @param executorService
+    *   \- Shared executor service to process the task in parallel
+    * @return
+    *   List of StartingPointWithSource and List of tasks for calculating starting points with usage in other classes
+    */
+  private def calculateStartingPoints(
+    sources: List[StoredNode],
+    executorService: ExecutorService
+  ): (List[StartingPointWithSource], List[UsageInput]) = {
+    val allExceptUsageInOtherClasses       = SourceStartingPointResultAggregator(sources.size)
+    val allExceptUsageInOtherClassesThread = new Thread(allExceptUsageInOtherClasses)
+    allExceptUsageInOtherClassesThread.setName("All except usage in other classes result aggregator")
+    allExceptUsageInOtherClassesThread.start()
+    sources.foreach(src =>
+      executorService.submit(new SourceToStartingPoints(src, allExceptUsageInOtherClasses.resultQueue))
+    )
+    allExceptUsageInOtherClassesThread.join()
```

Comment:
Is the order of results obtained here always stable or is it an arbitrary order? If so, I would suggest that we order the result such that it is returned in the original order of `sources`.

Reply:
final sorting on the result based on the original order of sources is done on combined results. That's how I could verify the consistency in results. - https://github.com/joernio/joern/pull/4429/files#diff-b887bf35647ceef08727f3fa1753c2cdcb473e966b53d598bbdf1b2f186dd6d4R40